cloud/aws/actuators/machine/actuator.go

```diff
@@ -188,6 +254,269 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	}
 }
 
+// Arguments:
+//  - annotation: The security group annotation.
+//  - src: The Kubernetes state.
+//  - dst: The AWS state.
+// We define them this way since we are going to change the dst (AWS) based on
+// the src (Kubernetes).
+//
+// Returns:
+//  - bool: Indicates if the state changed.
+//  - []*string{}: Array of securityGroupIDs, these should be passed to AWS
+//    if the state changed.
+//  - map[string]interface{}: The new annotation to set on the machine.
+//
+// First, we are comparing the annotation to the src to see if any security
+// groups were removed from the src since the last time Update was called.
+//
+// Based on this we create a new map[string]bool, representing groupID=>present.
+// groupID=>true indicates a new or already existing groupID.
+// groupID=>false indicates a groupID that was removed since last time we were
+// called.
+//
+// Once we have the above map representing the changes since we last ran, we
+// start to prepare a map, representing what we'd like state in AWS to look
+// like.
+// We first add the groupIDs from our above state that need to be kept.
+// We then compare the state in AWS to our above state.
+//  - If we find a groupID in AWS that isn't in our state, that is a security
+// 	  group that we don't manage. We add it to our new map with no further
+//    checking.
+//  - If we find a groupID in AWS that's in our state and flagged to be kept,
+//    we add it to the map.
+//  - If we find a groupID in AWS that's in our state and flagged for deletion,
+//    we do not add it to the map and we allow it to drop.
+//
+// Next we compare the map we have made after comparing our desired state and
+// the AWS state against AWS again to decide if any changes actually happened.
+//
+// Finally, we return three items listed at the top of this comment.
+// The caller then acts upon this information.
+func (a *Actuator) securityGroupsChanged(annotation map[string]interface{}, src []v1alpha1.AWSResourceReference, dst map[string]string) (bool, []*string, map[string]interface{}) {
```

Comment:
@chuckha @vincepri thoughts on the approach here? The lack of set theory primitives in Go drives me batty sometimes and I haven't internalized the idiomatic way of handling it yet.

Reply:
I think this is mostly personal preference. Usually I prefer using `map[string]struct{}`. The empty struct takes 0 bytes and checking if an item is in the map is as easy as `if _, ok := set["test"]`. The actual value in the map can be ignored at that point, which I find nice.

Reply:
Good point, I'd forgotten about the empty struct. I'll make that change where appropriate.

Reply:
Where possible, maps now use an empty struct if they don't care about the map value.